activesupport/lib/active_support/current_attributes.rb

```diff
@@ -200,8 +211,21 @@ def set(attributes, &block)
     # Reset all attributes. Should be called before and after actions, when used as a per-request singleton.
     def reset
       run_callbacks :reset do
-        self.attributes = {}
+        self.attributes = merge_defaults!({})
       end
     end
+
+    private
+      def merge_defaults!(attributes)
+        defaults.each_with_object(attributes) do |(name, default), values|
+          value =
+            case default
+            when Proc then default.call
+            else default.dup
+            end
```

Comment:
To compare to prior art, this is the code that Active Model uses when determining the default:

https://github.com/rails/rails/blob/beb6214bb00576073c8b47ee2adc436c5bcf098b/activemodel/lib/active_model/attribute/user_provided_default.rb#L17-L23